### DISCUSSION TITLE
Sim I/O or console data to-and-from external JS process, in real time

### DISCUSSION THREAD

**Discussion by toysonlocation** [microsoft/jacdac#969] [General]
After asking around on the regular makecode lists, I've been directed here in case someone hopefully knows how to achieve this: pipe makecode/microbit sim data in **real time** to another JS workflow (via ANY method, such as classic JS console or maybe a socket), and then have that other workflow also send back data (via simulated sensors, for example) to the makecode sim. The other workflow would be a real time processor of the sim data, written also in JavaScript and perhaps in another HTML frame, and pass back state for the simulated microbit to handle. Please advise!

**Comment by pelikhan** [comment]
Jacdac uses a mechanism called "control simmessages" in pxt that allows to send a buffer through a C++ function or through a HTML window message (to/from jacdac simulator iframe).

The API is implemented in pxt however loading iframes in the editor is restricted to known URLs, so you would have to create a local build of the editor to allow yours.

That being said, it may turn out that Jacdac is the right abstraction for you and we can make it work for your scenario. We already support a switching between simulated or hardware data in the micro:bit editor. I'm wondering what's the gap to make it useful in your scenario.

(Note I'm confused about the difference of both workflows)

**Reply by toysonlocation** [reply]
Thanks so much for the reply. I'm attaching a simple diagram of what we're after. Basically it's a virtual, continuous feed from the Car Simulator into the MakeCode "sensor" input which can be processed into a control loop by sending virtual motor/servo updates to a virtual car back in the Car Simulator. These are the biggest questions I have:

- How get real time data to flow to/from the Car Simulator
- How to present the Car Simulator virtual sensors as Sensor input in MakeCode

Ideally I wouldn't have to be maintaining a branch of MakeCode and microbit, but I'm open to that if it's necessary.
![control_flow_diagram](https://user-images.githubusercontent.com/39204807/164055957-81b1085f-8e4c-470d-b37e-4432eac9d828.png)

**Reply by pelikhan** [reply]
So the makecode editor being https will only talk to another https Web site; is the car simulator a hosted web app?

**Reply by toysonlocation** [reply]
I can certainly make it so if that's a requirement. Now when you say "site", and from you question, it sounds like you mean the makecode editor would have to post to a server. If so, how would our server "post back" real time sensor data to makecode? Or, is it always request-response/polling from makecode?

If on the other hand you mean it has to source from an HTTPS site, then are we talking shared JS objects/callbacks or maybe events?

**Reply by pelikhan** [reply]
Here is a potential way to do this without having to fork makecode. It is using jacdac (disclaimer).

This solution is based on the jacdac devtools, that allow to load a makecode editor side by side with another web site that's running the jacdac code. Both communicate via iframe messages.
You would load a custom web site that would then communicate to your robot simulator tool.

   makecode editor <-> proxy web site <-> robot simulator

Now messages are jacdac packets so you'll need to use use jacdac services to send/receive state updates between the makecode simulator and the robot simulator.

Concretely, the students would model their rover as a device with 2 motor + sonar services using jacdac blocks. Motor commands are sent from makecode to the motor simulator and sonar events are polled back from the motor simulator to makecode.

**Reply by pelikhan** [reply]
I wouldn't be against pro typing this with you; I can help with the jacdac plumbing but I am not well versed in robot simulators.

**Reply by toysonlocation** [reply]
Hi pelikhan - I'm super sorry for vanishing. I got covid in early May and am only now getting back on track for this project. I am also not well versed in robot simulators! However, I've done a number of visual/mechanical applications, and this is going to be part of a STEM curriculum that I'm pretty excited about. I really appreciate your (back in April) above note. What would you like to see on deciding whether or not this is something for you?

**Reply by pelikhan** [reply]
Please take a look at what we've built with jacdac. I don't have the time to take on something else at the moment.

If you can map your system as a "service", you could leverage our infrastructure.

**Reply by toysonlocation** [reply]
Thanks - I'll have a look at your most recent release and docs and probably have to inquire here and there.
